File: behavioral/strategy/ruby/lib/paypal_payment_strategy.rb
require_relative 'payment_strategy'

class PaypalPaymentStrategy < PaymentStrategy
  attr_reader :email

  def initialize(email)
    @email = email
  end

  # Overrides PaymentStrategy#pay
  def pay(amount)
    # Put verification or authorization logic here
    puts "Paying #{amount} via Paypal with email #{email}"
    true
  end
end

File: behavioral/strategy/java/src/PaypalPaymentStrategy.java
public class PaypalPaymentStrategy implements PaymentStrategy {
  private String email;

  public PaypalPaymentStrategy(String email) {
    this.email = email;
  }

  @Override
  public boolean pay(double amount) {
    // can have logic around email verification
    // checking balance
    // bascially implementation of the algorithm

    System.out.println("Paying " + amount + " via Paypal with email " + email);
    return true;
  }
}
